pkg/apis/metal3/v1alpha1/baremetalhost_types.go

```diff
@@ -193,6 +196,70 @@ type BMCDetails struct {
+type HardwareRAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.
+	// If unspecified or set be 0, the maximum capacity of disk will be used for logical disk.
+	// +kubebuilder:validation:Minimum=0
+	SizeGibibytes *int `json:"sizeGibibytes,omitempty"`
+
+	// RAID level for the logical disk. The following levels are supported: 0;1;1+0.
+	// +kubebuilder:validation:Enum="0";"1";"1+0"
+	Level string `json:"level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+
+	// Select disks with only rotational or solid-state storage
+	Rotational *bool `json:"rotational,omitempty"`
+
+	// Integer, number of disks to use for the logical disk. Defaults to minimum number of disks required
+	// for the particular RAID level.
+	// +kubebuilder:validation:Minimum=1
+	NumberOfPhysicalDisks *int `json:"numberOfPhysicalDisks,omitempty"`
+}
+
+// SoftwareRAIDVolume defines the desired configuration of volume in software RAID
+type SoftwareRAIDVolume struct {
+	// Size (Integer) of the logical disk to be created in GiB.
+	// If unspecified or set be 0, the maximum capacity of disk will be used for logical disk.
+	// +kubebuilder:validation:Minimum=0
+	SizeGibibytes *int `json:"sizeGibibytes,omitempty"`
+
+	// RAID level for the logical disk. The following levels are supported: 0;1;1+0.
+	// +kubebuilder:validation:Enum="0";"1";"1+0"
+	Level string `json:"level" required:"true"`
+
+	// Name of the volume. Should be unique within the Node. If not specified, volume name will be auto-generated.
+	// +kubebuilder:validation:MaxLength=64
+	Name string `json:"name,omitempty"`
+
+	// A list of device hints, the number of item must be greater than or equal to 2.
+	// +kubebuilder:validation:MinItems=2
+	PhysicalDisks []RootDeviceHints `json:"physicalDisks,omitempty"`
+}
+
+// RAIDConfig contains the configuration that are required to config RAID in Bare Metal server
+type RAIDConfig struct {
+	// If true will remove existing raid configuration and set new raid configuration.
+	DeleteExisting bool `json:"deleteExisting,omitempty"`
+
+	// The list of logical disks for hardware RAID, first volume is root volume
+	HardwareRAIDVolumes []HardwareRAIDVolume `json:"hardwareRAIDVolumes,omitempty"`
+
+	// The list of logical disks for software RAID, first volume is root volume,
+	// if HardwareRAIDVolumes is setted this items will be invalid.
+	// The number of created Software RAID devices must be 1 or 2.
+	// If there is only one Software RAID device, it has to be a RAID-1.
+	// If there are two, the first one has to be a RAID-1, while the RAID level for the second one can 0, 1, or 1+0.
+	// As the first RAID device will be the deployment device,
+	// enforcing a RAID-1 reduces the risk of ending up with a non-booting node in case of a disk failure.
+	// +kubebuilder:validation:MinItems=1
```

Comment:
This minimum seems wrong. It implies that we always require software RAID?

Reply:
Thanks, done.